lock-api-objects/src/main/java/com/palantir/lock/watch/ClientLockWatchEventLogImpl.java

```diff
@@ -0,0 +1,150 @@
+    private IdentifiedVersion identifiedVersion;
+    private LockWatchStateUpdate.Snapshot seed = failedSnapshot(UUID.randomUUID());
+
+    private ClientLockWatchEventLogImpl() {
+        identifiedVersion = IdentifiedVersion.of(UUID.randomUUID(), Optional.empty());
+        eventLog = new ConcurrentSkipListMap<>();
+    }
+
+    @Override
+    public IdentifiedVersion getLatestKnownVersion() {
+        return identifiedVersion;
+    }
+
+    @Override
+    public IdentifiedVersion processUpdate(LockWatchStateUpdate update) {
+        if (update.logId().equals(identifiedVersion.id())) {
+            update.accept(processingVisitor);
+            return identifiedVersion;
+        } else {
+            update.accept(newLeaderVisitor);
+            return identifiedVersion;
+        }
+    }
+
+    @Override
+    public TransactionsLockWatchEvents getEventsForTransactions(
+            Map<Long, Long> timestampToVersion,
+            IdentifiedVersion version) {
+        if (!version.id().equals(identifiedVersion.id())) {
+            return TransactionsLockWatchEvents.failure(seed);
+        }
+
+        Long latestVersion = Collections.max(timestampToVersion.values());
+        ConcurrentSkipListMap<Long, LockWatchEvent> logSnapshot = eventLog.clone();
+
+        if (logSnapshot.isEmpty()) {
+            return TransactionsLockWatchEvents.success(ImmutableList.of(), timestampToVersion);
+        }
+
+        long oldestVersion = version.version().orElseGet(logSnapshot::firstKey);
+        Preconditions.checkArgument(oldestVersion <= latestVersion, "startVersion should be before endVersion");
+
+        long startKey = logSnapshot.ceilingKey(oldestVersion);
+        long endKey = logSnapshot.floorKey(latestVersion);
+
+        return TransactionsLockWatchEvents.success(
+                new ArrayList<>(logSnapshot.subMap(startKey, endKey).values()),
+                timestampToVersion);
+    }
+
+    private synchronized void processSuccess(LockWatchStateUpdate.Success success) {
+        setToLatestVersion(IdentifiedVersion.of(success.logId(), Optional.of(success.lastKnownVersion())));
+        success.events().forEach(event -> eventLog.put(event.sequence(), event));
+    }
+
+    private synchronized void processSnapshot(LockWatchStateUpdate.Snapshot snapshot) {
+        setToLatestVersion(IdentifiedVersion.of(snapshot.logId(), Optional.of(snapshot.lastKnownVersion())));
+        eventLog.clear();
+        seed = snapshot;
+    }
+
+    private synchronized void processFailed(LockWatchStateUpdate.Failed failed) {
+        setToLatestVersion(IdentifiedVersion.of(failed.logId(), Optional.empty()));
+        eventLog.clear();
+        seed = failedSnapshot(failed.logId());
+    }
+    
+    private static LockWatchStateUpdate.Snapshot failedSnapshot(UUID uuid) {
+        return LockWatchStateUpdate.snapshot(uuid, -1L, ImmutableSet.of(), ImmutableSet.of());
+    }
+
+    private void setToLatestVersion(IdentifiedVersion newVersion) {
```

Comment:
This may be overly cautious, but I can imagine a case where a successful update A is actually done after B, even though B covers A and goes beyond, i.e.

A: |----|
B: |----------|

Don't want the version to go backwards on the same leader ID.